lib/Providers/OpenIDConnectProvider.php

```diff
@@ -0,0 +1,186 @@
+     * @return array associative array of claims decoded from the id token
+     */
+    public function verifyIdToken($id_token)
+    {
+        try {
+            $keys = $this->getSigningKeys();
+            $claims = JWT\JWT::decode($id_token, $keys, ['RS256']);
+            if ($claims->aud !== $this->clientId) {
+                throw new IdentityProviderException("ID token has incorrect audience", 0, $claims->aud);
+            }
+            if ($claims->iss !== $this->issuer) {
+                throw new IdentityProviderException("ID token has incorrect issuer", 0, $claims->iss);
+            }
+        } catch (\UnexpectedValueException $e) {
+            throw new IdentityProviderException("ID token validation failed", 0, $e->getMessage());
+        }
+    }
+
+    protected function prepareAccessTokenResponse(array $result)
+    {
+        $result = parent::prepareAccessTokenResponse($result);
+        $this->verifyIdToken($result['id_token']);
+        return $result;
+    }
+
+    protected function getOpenIDConfiguration() {
+        if (isset($this->openIdConfiguration)) {
+            return $this->openIdConfiguration;
+        }
+
+        $config = $this->getParsedResponse($this->getRequest('GET', $this->issuer . self::CONFIGURATION_PATH));
+        $this->openIdConfiguration = $config;
+        return $config;
+    }
+
+    public static function base64url_decode($input) {
+        return base64_decode(strtr($input, '-_', '+/'));
+    }
+
+    protected function getSigningKeys() {
+        $url = $this->getOpenIDConfiguration()['jwks_uri'];
+        $jwks = $this->getParsedResponse($this->getRequest('GET', $url));
+        $keys = [];
+        foreach ($jwks['keys'] as $key) {
+            $kid = $key['kid'];
+            if (array_key_exists('x5c', $key)) {
+                $x5c = $key['x5c'];
+                $keys[$kid] = "-----BEGIN CERTIFICATE-----\n" . $x5c[0] . "\n-----END CERTIFICATE-----";
+            } else if ($key['kty'] === 'RSA') {
+                $e = self::base64url_decode($key['e']);
+                $n = self::base64url_decode($key['n']);
+                $keys[$kid] = \RobRichards\XMLSecLibs\XMLSecurityKey::convertRSA($n, $e);
+            } else {
+                error_log("Failed to load key data for key id: " . $kid);
+            }
+        }
+        return $keys;
+    }
+
+    /**
+     * Returns the base URL for authorizing a client.
+     *
+     * Eg. https://oauth.service.com/authorize
+     *
+     * @return string
+     */
+    public function getBaseAuthorizationUrl()
+    {
+        return $this->getOpenIDConfiguration()["authorization_endpoint"];
+    }
+
+    /**
+     * Returns the base URL for requesting an access token.
+     *
+     * Eg. https://oauth.service.com/token
+     *
+     * @param array $params
+     * @return string
+     */
+    public function getBaseAccessTokenUrl(array $params)
+    {
+        return $this->getOpenIDConfiguration()["token_endpoint"];
```

Comment:
It's also a bit weird that `$params` here and `$token` in the next one aren't even used. What happens if `token_endpoint` is missing from the configuration, btw? Is that even possible?

Reply:
These parameters are also defined in the oauth2-client library's api. No idea what they are supposed to be used for, but we can't just remove them here. As for missing the endpoint in the configuration, that would require a OP that doesn't comply with the spec. We should probably check that we have a minimum required config in getOpenODConfiguration() and give some sort of sensible feedback to the user if not.

Reply:
I think that's a sensible idea for the `getOpenIDConfiguration()` method, indeed! At least, we should not take things for granted, just in case.